distributed/client.py

```diff
@@ -727,10 +732,42 @@ def __init__(
 
         ReplayExceptionClient(self)
 
+    @contextmanager
+    def as_current(self):
+        """Thread-local, Task-local context manager that causes the Client.current class
+        method to return self. This is used when a method of Client needs to propagate a
+        reference to self deep into the stack through generic methods that shouldn't be
+        aware of this class.
+        """
+        # Python 3.6; contextvars are thread-local but not Task-local.
+        # We can still detect a race condition.
+        if sys.version_info < (3, 7) and _current_client.get() not in (self, None):
+            raise RuntimeError(
+                "Detected race condition where get_dataset() is invoked in "
```

Comment:
This error message still mentions `get_dataset`, should probably be made a bit more generic.

Reply:
Fixed